staging/src/k8s.io/apiserver/pkg/util/flowcontrol/reqmgmt.go

```diff
@@ -115,66 +110,105 @@ type requestManagementSystem struct {
+	queueSetFactory fq.QueueSetFactory,
 	serverConcurrencyLimit int,
 	requestWaitLimit time.Duration,
-	clk clock.Clock,
-) Interface {
+	preservingFlowSchemas []*rmtypesv1alpha1.FlowSchema,
+	preservingPriorityLevels []*rmtypesv1alpha1.PriorityLevelConfiguration,
+) *requestManagementSystem {
 	reqMgmt := &requestManagementSystem{
-		clk:                    clk,
-		configQueue:            workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 8*time.Hour), "req_mgmt_config_queue"),
-		plLister:               informerFactory.Flowcontrol().V1alpha1().PriorityLevelConfigurations().Lister(),
-		fsLister:               informerFactory.Flowcontrol().V1alpha1().FlowSchemas().Lister(),
+		queueSetFactory:        queueSetFactory,
 		serverConcurrencyLimit: serverConcurrencyLimit,
 		requestWaitLimit:       requestWaitLimit,
+		flowcontrolClient:      flowcontrolClient,
+	}
+	klog.V(2).Infof("NewRequestManagementSystem with serverConcurrencyLimit=%d, requestWaitLimit=%s", serverConcurrencyLimit, requestWaitLimit)
+	reqMgmt.initializeConfigController(informerFactory)
+	emptyRMState := &requestManagementState{
+		priorityLevelStates: make(map[string]*priorityLevelState),
+	}
+	reqMgmt.curState.Store(emptyRMState)
+	if len(preservingFlowSchemas) > 0 || len(preservingPriorityLevels) > 0 {
+		reqMgmt.digestConfigObjects(
+			preservingPriorityLevels,
+			preservingFlowSchemas,
+		)
 	}
-	// TODO: finish implementation
 	return reqMgmt
 }
 
-// Run runs the controller that deals with config API objects
-func (reqMgmt *requestManagementSystem) Run(stopCh <-chan struct{}) error {
-	// TODO: implement
-	return nil
-}
-
 // RequestDigest holds necessary info from request for flow-control
 type RequestDigest struct {
 	RequestInfo *request.RequestInfo
 	User        user.Info
 }
 
-func (reqMgmt *requestManagementSystem) Wait(requestDigest RequestDigest) (execute bool, afterExecute func()) {
+func (reqMgmt *requestManagementSystem) Wait(requestDigest RequestDigest) (bool, func()) {
 	for {
-		rmState := reqMgmt.curState.Load().(*RequestManagementState)
+		rmState := reqMgmt.curState.Load().(*requestManagementState)
+
+		var matchingflowSchemaName string
+		var matchingpriorityLevelName string
+		var matchingFlowDistinguisherMethod *rmtypesv1alpha1.FlowDistinguisherMethod
+
+		// 1. computing flow
 		fs := rmState.pickFlowSchema(requestDigest)
-		// RequestManagementState is constructed to guarantee that the following succeeds
-		ps := rmState.priorityLevelStates[fs.Spec.PriorityLevelConfiguration.Name]
+		_, defaultPLExists := rmState.priorityLevelStates[rmtypesv1alpha1.PriorityLevelConfigurationNameDefault]
+		switch {
+		case fs != nil: // successfully matched a flow-schema
+			matchingflowSchemaName = fs.Name
+			matchingpriorityLevelName = fs.Spec.PriorityLevelConfiguration.Name
+			matchingFlowDistinguisherMethod = fs.Spec.DistinguisherMethod
+		case defaultPLExists: // falling to global-default
```

Comment:
The KEP says that if the request does not match any actually existing FlowSchema API object then:
(1) if the request's authenticated groups include "system:masters" then the request is exempt and handled accordingly,
(2) otherwise the request is handled as if it matched the default "workload-low" FlowSchema, albeit with a possibly different name, and mapped to a priority level equivalent to the one here called "default" (regardless of whether an API object for it exists or not).

The logic here does not address case (1).  For case (2) the flow distinguisher method should be by namespace rather than user.

Reply:
refactored